### PR TITLE
Fix Template Rendering

### LINKED ISSUE #21311: Showing fields from M2M translations error
### Describe the Bug

I include a translation field on a M2M collection . When I show on a M2M field it shows the template for translation instead the content field.
 

### To Reproduce

Collections:
- Department
- Employees

1. Create a M2M relationship field between Department and Employees, on Department: This will create the Department_employees collection.
2. Create a translation field on department_employees table. This vill create a department_employees_translation table
3. Modify M2M field interface adding columns:
   - translations -> Job
   - translations -> Skill
4. Entering skill1 on translation en-USA skill field, and job1 on translation en-USA job field
5. On department layout, job and skill field show correct values
dept name job skill
IT John Clark - job1 - skill1  
6. On department edition page, job and skill fields on M2M field list, show the same value 
John Clark - job1skill1 - job1skill1
They are showing template translation field values instead of independent one

https://github.com/directus/directus/assets/108670131/8cf8615c-0441-4246-b619-fa601085cb48

 

### Directus Version

v10.8.3

### Hosting Strategy

Self-Hosted (Docker Image)

### LINKED ISSUE #22734: Many-to-Many (m2m) fields to not render translations properly without saving
### Describe the Bug

In the the following schema I have the collection 'Test Collection', which only consists of an Many-To-Many (m2m) interface with another collection ('Test Collection 2').
'Test Collection 2' consists of a translation interface, which only has a single input interface.
A simple visualization of this looks like this. 
grafik

I add a new item via the m2m Interface "Create New" button like this:
Screenshot 2024-06-12 092352

Since my user language is English and I only provided one English translation, I would assume that the M2M interface displays the entered translation item in an appropiate way.
However, the M2M interface displays the raw translation create,update,delete-object.
Screenshot 2024-06-12 092405

Only after saving the item, which I'm currently editing in 'Test Collection', the M2M interface displays the translation properly: 
Screenshot 2024-06-12 092416

This occurs even though I managed the display styles properly. So that the translation interface display is set to the following:
Screenshot 2024-06-12 092435

And the interface/ display of the M2M interface is set to 'Translations' of the 'Test Collection 2'.
Screenshot 2024-06-12 092522

**Expected Behavior:**
This behavior is confusing for users entering items in this m2m interface.
Since it is hard to track which values have already been entered without saving and reloading the page.
I would expect that the translation is already rendered properly, without having to save the page of 'Test Collection'.

### To Reproduce

1. Create two empty collections
2. The first collection contains a many-to-many (m2m) interface to the second collection
3. The second collection contains a translations interface with any fields (i.e., an input field)
4. Set the display style of the translations field to 'Translation' and set the wished fields to be shown in the display style
5. Set the interface style of the m2m interface to 'Related Values' and set it to the 'Translations' field of the second collection.
6. Go to the first collection and add a new item of the second collection with a translated fields to the m2m interface ("Create New" Button).
7. **Issue:** The new entry in the m2m interface is not rendered properly. It shows a raw string containing: {"create" : { . . .}, "update": { . . .}, "delete": { . . .}}, which describes the translation.
8. Save the page for the first collection and the translation value in the m2m interface is rendered properly.

### Directus Version

v.10.12.1

### Hosting Strategy

Self-Hosted (Docker Image)

### Database

PostgreSQL 13.15

### LINKED ISSUE #22788: Subtitle Link Undefined
### Describe the Bug

When setting a related field to the subtitle part of a card, it doesn't show the related string and the link leads to an undefined 404. 

image
image

If you set the subtitle to the related items ID, the list is blank, but the link takes you to the correct item.

### To Reproduce

Go to the card view and add a related field to the subtitle and click to attempt to see it's name.

### Directus Version

10.11.2

### Hosting Strategy

Directus Cloud

### Database

_No response_

### PR DESCRIPTION
**Pull request by connorwinston**
## Scope

What's changed:

- `render-template` now supports rendering lists of nested fields and their displays.
- Fix Label Display not listing labels.
- Fix Formatted Value Display not joining array of strings with `, `

## Potential Risks / Drawbacks

- Multiple nested relationships could get complicated

## Review Notes / Questions

- Need to test all displays
- Find more edge cases

---

Fixes #22788

### CODE REVIEW

File: app/src/views/private/components/render-template.vue [hunk 1]
```diff
@@ -27,140 +27,121 @@ const templateEl = ref<HTMLElement>();
 
 const regex = /({{.*?}})/g;
 
+const getNestedValues = (data: any, path: string) => {
+	const pathParts = path.split('.');
+	let currentData = data;
+
+	for (const part of pathParts) {
+		if (Array.isArray(currentData)) {
+			currentData = flatMap(currentData.map((item) => get(item, part)));
```
Comment (formfcw) [inline app/src/views/private/components/render-template.vue:36]: Can we replace `flatMap` with the native `.flatMap()`? Is it functionally the same?

File: app/src/views/private/components/render-template.vue [hunk 2]
```diff
@@ -27,140 +27,121 @@ const templateEl = ref<HTMLElement>();
...
-		collection: field.collection,
-		field: field.field,
-	};
-}
 </script>
 
 <template>
 	<div ref="templateEl" class="render-template">
 		<span class="vertical-aligner" />
 		<template v-for="(part, index) in parts" :key="index">
```
Comment (formfcw) [inline app/src/views/private/components/render-template.vue:116]: We should avoid using array indices for the `:key` attribute.

File: app/src/views/private/components/render-template.vue [hunk 3]
```diff
@@ -27,140 +27,121 @@ const templateEl = ref<HTMLElement>();
...
-					:field="part.field"
-				/>
-
-				<template #fallback>
-					<span>{{ part.value }}</span>
-				</template>
-			</v-error-boundary>
-			<span v-else-if="typeof part === 'string'" :dir="direction">{{ translate(part) }}</span>
-			<span v-else>{{ part }}</span>
+			<template v-for="(subPart, subIndex) in part" :key="subIndex">
```
Comment (formfcw) [inline app/src/views/private/components/render-template.vue:117]: Optional: Can we find more descriptive names than parts, part and subPart?

File: app/src/views/private/components/render-template.vue [hunk 4]
```diff
@@ -27,140 +27,121 @@ const templateEl = ref<HTMLElement>();
...
+
+			// No need to render the empty display overhead in this case
+			if (component === 'raw') return [value];
+
+			const displayInfo = useExtension(
+				'display',
+				computed(() => component ?? null),
+			);
+
+			if (!displayInfo.value) return [value];
```
Comment (formfcw) [inline app/src/views/private/components/render-template.vue:76]: Just for my understanding:
Why do we return `[value]` instead of `value` since `getNestedValues()` already returns an array? Then below if there are displays, we return an array of objects, yes? And at the bottom we map the values in `.map((p) => p ?? null),`, where `p` will always be an array, so it can't be `null`, right? And finally in the template, we have two loops, because we are wrapping these values in an array, right?
Does that make sense?

**Review by formfcw — CHANGES_REQUESTED** [review]
Great job @connorwinston 🚀 

However, there is one thing that immediately caught my eye:

When displaying an `O2M → title` I would expect to see a comma separated list …

cards

… like in the table layout:

table

File: app/src/views/private/components/render-template.vue (hunk 1 above)
Comment (connorwinston) [reply app/src/views/private/components/render-template.vue:36]: We could. The lodash implementation is more performant since it it is not recursive where the native one is recursive.

File: app/src/views/private/components/render-template.vue (hunk 4 above)
Comment (connorwinston) [reply app/src/views/private/components/render-template.vue:76]: Good catch!
```suggestion
			if (!displayInfo.value) return value;
```

File: app/src/views/private/components/render-template.vue (hunk 3 above)
Comment (connorwinston) [reply app/src/views/private/components/render-template.vue:117]: Like templateParts, templatePart, and templateSubPart?

File: app/src/views/private/components/render-template.vue (hunk 2 above)
Comment (connorwinston) [reply app/src/views/private/components/render-template.vue:116]: There really isn't a unique identifier for each of the parts of the array. We can't use the value because those are not guaranteed to be unique. Thoughts on how to avoid indice?

File: app/src/views/private/components/render-template.vue (hunk 1 above)
Comment (formfcw) [reply app/src/views/private/components/render-template.vue:36]: As far as I found out, `.flatMap()` is not recursive and flattens only one level deep — see the first line in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap. Since the lodash version is not used throughout the project, I would recommend using the native one.

File: app/src/views/private/components/render-template.vue (hunk 2 above)
Comment (formfcw) [reply app/src/views/private/components/render-template.vue:116]: Well, I personally would not use any `:key` in this scenario, but since `:key="index"` is used quite often throughout the project, let's be consistent and go with your implementation.

File: app/src/displays/formatted-value/formatted-value.vue [hunk 5]
```diff
@@ -96,7 +96,7 @@ const displayValue = computed(() => {
 
 	if (isNil(props.value) || props.value === '') return null;
 
-	let value = String(props.value);
+	let value = Array.isArray(props.value) ? props.value.join(', ') : String(props.value);
```
Comment (formfcw) [inline app/src/displays/formatted-value/formatted-value.vue:99]: To prevent strings like `, Item 2, , `, we need to ensure that each array element exists, like `props.value.map((p) => p || p === 0 ? p : '--').join(', ')`. Ideally we should use the `<value-null>` component instead of `'--'`, but I’ll leave that up to you.

**Review by formfcw — DISMISSED** [review]
Thank you @connorwinston :rocket: 

A few other things I found while testing that you might want to take a look at.

File: app/src/views/private/components/render-template.vue [hunk 6]
```diff
@@ -27,140 +27,121 @@ const templateEl = ref<HTMLElement>();
...
-		collection: field.collection,
-		field: field.field,
-	};
-}
 </script>
 
 <template>
 	<div ref="templateEl" class="render-template">
 		<span class="vertical-aligner" />
 		<template v-for="(part, index) in parts" :key="index">
```
[the remark this answers is not in the document]
Comment (connorwinston) [reply app/src/views/private/components/render-template.vue:116]: Discussed in meeting about this and resolving it.

File: app/src/displays/formatted-value/formatted-value.vue (hunk 5 above)
Comment (connorwinston) [reply app/src/displays/formatted-value/formatted-value.vue:99]:
```suggestion
	let value = Array.isArray(props.value) ? props.value.map((v) =>isNil(v) || v === '' ? '--' : v).join(', ') : String(props.value);
```

**Review by AlexGaillard — APPROVED** [review]
NIce addition. Thanks @connorwinston 🚀
